dark-souls-class.js

```diff
@@ -5,33 +5,64 @@ const readline = require('readline');
 const rl = readline.createInterface({
     input: process.stdin,
     output: process.stdout
-  });
+});
+
+class Inventory {
+    constructor(_items = []) {
+        this.items = _items;
+    }
+
+    /**
+     * Add items to the inventory
+     * @param {array} items
+     * @returns {*[]}
+     */
+    addItems(items) {
+        this.items = this.items.concat(items);
+        return this.items;
+    }
+
+    /**
+     * Remove an item from the inventory
+     * @param {string} name
+     */
+    removeItem(name) {
+        this.items = this.items.filter(item => item.name !== name);
+        return this.items;
+    }
+}
 
 class Character {
-    constructor(name, sex, gift, physique, stats, inventory) {
+    constructor(name, sex, gift, physique, stats, inventory = new Inventory()) {
         this.name = name;
         this.sex = sex;
         this.gift = gift;
         this.physique = physique;
-        this.inventory = {
-                        'weapon': null,
-                        'shield': null,
-                        'magic': null,
-                        'other': null
-                        }
-        this.stats = [ 
-                    'level', 
-                    'vitality',
-                    'attunement', 
-                    'endurance', 
-                    'strength', 
-                    'dexterity', 
-                    'resistance', 
-                    'intellegence', 
-                    'faith'
-                    ] 
- 
+        this.inventory = inventory;
+        this._stats = stats;
     }
+
+    // setters => change (mutate) properties, with validation
+    set stats(newStats) {
```

Comment:
The setter doesn't seem to work. Try running the following tests

```
let guy = new Knight();
guy.level = '9';
//should throw Error

log(guy.stats)
log(guy.level)
log(guy.stats.level)
```

Reply:
The setter is accessed through `guy.stats`. To level up you would do something like

`guy.stats = {
    level = 9,
}`

The `guy.stats` is also the getter, I.e. `console.log(guy.stats.level);`.

The refactor doesn't currently place each stat directly on the `Character`, but could easily be added by doing something like this in the `Knight` constructor

`for (let key in this.stats) {
    this[key] = this.stats[key];
}`

the setter doesn't currently support this though

Reply:
`guy.stats = { level: 9 }
`
Adds stats

Why does `Character` need the stats? Won't they be different for each new character?

Reply:
They are different. Take the following:

`const warrior = new Warrior('Q', 'Male', 'Twin Humanities', 'Thin');

console.log(warrior.stats);
// Initial stats, unique to Warrior class:
// {
//     level: 4,
//     vitality: 11,
//     attunement: 8,
//     endurance: 12,
//     strength: 13,
//     dexterity: 13,
//     resistance: 11,
//     intelligence: 9,
//     faith: 9
// }

// We only need to update the stats that are necessary
const newWarriorStats = {
    level: 10,
}

// Setter, not obvious that this is a function, therefore
// hiding implementation details from the consumer
warrior.stats = newWarriorStats;

console.log(warrior.stats);
// Updated stats, only level has changed:
// {
//     level: 10,
//     vitality: 11,
//     attunement: 8,
//     endurance: 12,
//     strength: 13,
//     dexterity: 13,
//     resistance: 11,
//     intelligence: 9,
//     faith: 9
// }`

This is different from Knight, confirmed by the following:

`const knight = new Knight('Q', 'Male', 'Twin Humanities', 'Thin');

console.log(knight.stats);
// Initial stats, unique to Knight:
// {
//     level: 5,
//     vitality: 14,
//     attunement: 10,
//     endurance: 10,
//     strength: 11,
//     dexterity: 11,
//     resistance: 10,
//     intelligence: 9,
//     faith: 11
// }

// We only need to update the stats that are necessary
const newKnightStats = {
    level: 9,
}

// Setter, not obvious that this is a function, therefore
// hiding implementation details from the consumer
knight.stats = newKnightStats;

console.log(knight.stats);
// Updated stats, only level has changed:
// {
//     level: 9,
//     vitality: 14,
//     attunement: 10,
//     endurance: 10,
//     strength: 11,
//     dexterity: 11,
//     resistance: 10,
//     intelligence: 9,
//     faith: 11
// }`

Each Dark Souls class is an extension of `Character`. This is the beauty of inheritance. I hope this makes it clear.